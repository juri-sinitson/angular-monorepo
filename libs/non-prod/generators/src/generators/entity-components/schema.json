{
  "$schema": "https://json-schema.org/schema",
  "$id": "EntityComponents",
  "title": "",
  "type": "object",
  "properties": {
    "entityName": {
      "type": "string",
      "description": "The name of the component to create.",
      "x-prompt": "What component name would you like to use?"
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "x-prompt": "What project name would you like to use?"
    },
    "crud": {
      "type": "boolean",
      "default": false,
      "description": "The name of the project."
    }
  },
  "required": ["entityName", "project"]
}
