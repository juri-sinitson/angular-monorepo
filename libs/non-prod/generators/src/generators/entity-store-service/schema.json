{
  "$schema": "https://json-schema.org/schema",
  "$id": "EntityStoreService",
  "title": "",
  "type": "object",
  "properties": {
    "entityName": {
      "type": "string",
      "description": "The name of the component to create.",
      "x-prompt": "What component name would you like to use?"
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "x-prompt": "What project name would you like to use?"
    }
  },
  "required": ["entityName", "project"]
}
