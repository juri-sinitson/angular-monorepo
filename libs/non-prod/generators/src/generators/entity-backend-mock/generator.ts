import {
  readProjectConfiguration,
  formatFiles,
  generateFiles,
  Tree,
} from '@nx/devkit';
import * as path from 'path';
import { strings } from '@angular-devkit/core';

import { EntityBackendMockGeneratorSchema } from './schema';

export async function entityBackendMockGenerator(
  tree: Tree,
  options: EntityBackendMockGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;
  const entityClassName = strings.classify(options.entityName);

  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, {
    entityInterfaceName: `${entityClassName}Interface`,
    entityClassName,
    entityInUrlNamePlural: `${strings.dasherize(options.entityName)}s`,
  });
  await formatFiles(tree);
}

export default entityBackendMockGenerator;
