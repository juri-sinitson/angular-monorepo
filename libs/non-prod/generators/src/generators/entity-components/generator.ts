import {
  formatFiles,
  generateFiles,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';

import { strings } from '@angular-devkit/core';

import * as path from 'path';
import { EntityComponentsGeneratorSchema } from './schema';

export async function entityComponentsGenerator(
  tree: Tree,
  options: EntityComponentsGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;

  const componentFolderName = `${strings.dasherize(options.entityName)}s`;
  const filesRoot = `${projectRoot}/src/lib/components/${componentFolderName}`;
  const entityName = strings.capitalize(options.entityName);
  generateFiles(tree, path.join(__dirname, 'files'), filesRoot, {
    className: `${strings.classify(options.entityName)}s`,
    fileName: `${strings.dasherize(options.entityName)}s`,
    entityName: strings.capitalize(options.entityName),
    entityInterfaceName: `${entityName}Interface`,
    selectorName: `${strings.dasherize(options.entityName)}s`,
    entityServiceClassName: `${strings.classify(options.entityName)}EntityStoreService`,
    entityServiceName: `${strings.camelize(options.entityName)}Service`,
  });

  await formatFiles(tree);
}

export default entityComponentsGenerator;
