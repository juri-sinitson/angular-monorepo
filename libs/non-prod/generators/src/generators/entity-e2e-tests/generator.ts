import {
  readProjectConfiguration,
  formatFiles,
  generateFiles,
  Tree,
} from '@nx/devkit';
import { strings } from '@angular-devkit/core';
import * as path from 'path';

import { EntityE2eTestsGeneratorSchema } from './schema';

export async function entityE2eTestsGenerator(
  tree: Tree,
  options: EntityE2eTestsGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;
  const entitySelectorName = `${strings.dasherize(options.entityName)}`;
  const entitySelectorNamePlural = `${entitySelectorName}s`;

  const className = `${strings.classify(options.entityName)}`;  

  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, {
    urlName: `${strings.camelize(options.entityName)}sUrl`,
    entitySelectorNamePlural,
    classNamePlural: `${className}s`,
  });
  await formatFiles(tree);
}

export default entityE2eTestsGenerator;
