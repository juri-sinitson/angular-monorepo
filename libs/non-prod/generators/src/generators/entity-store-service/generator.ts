import {
  readProjectConfiguration,
  formatFiles,
  generateFiles,
  Tree,
} from '@nx/devkit';
import { strings } from '@angular-devkit/core';
import * as path from 'path';

import { EntityStoreServiceGeneratorSchema } from './schema';

export async function entityStoreServiceGenerator(
  tree: Tree,
  options: EntityStoreServiceGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;
  
  const filesRoot = `${projectRoot}/src/lib/stores`;
  const className = `${strings.classify(options.entityName)}`;
  const classNamePlural = `${className}s`;
  const fileName = `${strings.dasherize(options.entityName)}`;

  generateFiles(tree, path.join(__dirname, 'files'), filesRoot, {
    className,
    fileName,    
    entityName: fileName,
    classNamePlural,
    getUrlName: `getAll${classNamePlural}Url`,
    interfaceName: `${strings.classify(options.entityName)}Interface`,    
  });
  await formatFiles(tree);
}

export default entityStoreServiceGenerator;
