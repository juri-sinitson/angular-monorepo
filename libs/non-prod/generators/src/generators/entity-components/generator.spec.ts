import { fail } from 'assert';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree, addProjectConfiguration } from '@nx/devkit';

import { entityComponentsGenerator } from './generator';
import { EntityComponentsGeneratorSchema } from './schema';

describe('entity-components generator', () => {
  let tree: Tree;
  const project = 'my-lib';
  const entityName = 'MyProduct';
  const fileNamePrefix = 'my-products';
  const root = `libs/${project}`;
  const fileRoot = `${root}/src/lib/components/my-products`;

  const presentationalComponentFile = `${fileRoot}/${fileNamePrefix}.component.ts`;
  const storyFile = `${fileRoot}/${fileNamePrefix}.component.stories.ts`;
  const smartComponentFile = `${fileRoot}/${fileNamePrefix}-smart.component.ts`;
  const crud = true;

  const options: EntityComponentsGeneratorSchema = {     
    entityName, 
    project,
    crud,
  };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should match snapshot ON any BY any', async () => {
    
    addProjectConfiguration(tree, project, { root });
    await entityComponentsGenerator(tree, options);    

    const presentationalComponentFileContent = tree.read(presentationalComponentFile);
    if(presentationalComponentFileContent) {      
      expect(presentationalComponentFileContent.toString('utf-8')).toMatchSnapshot();
    } else {
      fail('Presentational component file not found in virtual tree!');
    }

    const storyFileFileContent = tree.read(storyFile);
    if(storyFileFileContent) {      
      expect(storyFileFileContent.toString('utf-8')).toMatchSnapshot();
    } else {
      fail('Component story file not found in virtual tree!');
    }

    const smartComponentFileContent = tree.read(smartComponentFile);
    if(smartComponentFileContent) {      
      expect(smartComponentFileContent.toString('utf-8')).toMatchSnapshot();
    } else {
      fail('Smart component file not found in virtual tree!');
    }
  });
});
