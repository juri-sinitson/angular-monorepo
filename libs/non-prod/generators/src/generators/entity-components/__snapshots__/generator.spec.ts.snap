// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity-components generator should match snapshot ON any BY any 1`] = `
"import { ChangeDetectionStrategy, Component } from '@angular/core';

import { DialogModule } from 'primeng/dialog';

import {
  BasicTableComponent,
  CommonWrapperComponent,
} from '@angular-monorepo/shared/ui-common';
import { AbstractEntitiesListComponent } from '@angular-monorepo/shared/ui-common';

// -- STEP 1: ADJUST THE ENTITY INTERFACE PATH! --
import { MyProductInterface as EntityInterface } from '@angular-monorepo/shared-business/examples';
import { MyProductFormComponent as EntityFormComponent } from './MyProduct-form.component';

@Component({
  selector: 'angular-monorepo-my-products',
  standalone: true,
  imports: [
    // PrimeNG
    DialogModule,

    // Own
    BasicTableComponent,
    CommonWrapperComponent,
    EntityFormComponent,
  ],
  template: \`
    <common-common-wrapper
      [messages]="messages()"
      [isLoading]="isListLoading()"
      [header]="header()"
      [noData]="noData()"
    >
      <common-basic-table
        [columns]="columns"
        [data]="data()"
        [crud]="crud()"
        [isError]="isError()"
        (onDelete)="deleteHandler($event)"
        (onEdit)="editHandler($event)"
        (onNew)="newHandler()"
      ></common-basic-table>
    </common-common-wrapper>
    <!-- Generator template if CRUD is enabled -->
    @if (crud()) {
    <p-dialog
      [visible]="showEntityDialog()"
      [style]="{ width: '450px' }"
      [modal]="true"
      [closable]="false"
      styleClass="p-fluid"
    >
      <ng-template pTemplate="content">
        <angular-monorepo-MyProduct-form
          [data]="selectedEntity()"
          [header]="'MyProduct'"
          [messages]="messages()"
          [isLoading]="isFormLoading()"
          (onSubmit)="submitHandler($event)"
          (onCancel)="cancelHandler()"
        >
        </angular-monorepo-MyProduct-form>
      </ng-template>
    </p-dialog>
    }
    <!-- Generator template -->
  \`,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MyProductsComponent extends AbstractEntitiesListComponent<EntityInterface> {
  protected override getColumns(): Array<[keyof EntityInterface, string]> {
    // -- STEP 2: ADJUST THE COLUMNS NAMES AND THE SEQUENCE!
    // Adjusting the columns names and the sequence
    return [
      ['name', 'Name'],
      ['category', 'Category'],
      ['description', 'Description'],
      ['inventoryStatus', 'Status'],
      ['code', 'Code'],
      ['price', 'Price'],
      ['quantity', 'Quantity'],
      ['rating', 'Rating'],
    ];
  }
}
"
`;

exports[`entity-components generator should match snapshot ON any BY any 2`] = `
"import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import {
  applicationConfig,
  type Meta,
  type StoryObj,
} from '@storybook/angular';

import { ConfirmationService } from 'primeng/api';
import { ConfirmDialogModule } from 'primeng/confirmdialog';

import { MyProductsComponent as EntitiesComponent } from './my-products.component';
// -- STEP 1: Adjust the path to the entity interface
import { MyProductInterface as EntityInterface } from '@angular-monorepo/shared-business/examples';
import {
  MessageInterface,
  commonAppConfig,
} from '@angular-monorepo/shared/util-common';

import {
  errorStory,
  getFirstElemByText,
  expectText,
  loadingStory,
  noDataStory,
  ConfirmNotImplementedWrapperComponent,
  primaryTableStory,
  getCanvas,
  expectNoElem,
} from '@angular-monorepo/shared/util-common-non-prod';

const entitiesList: EntityInterface[] =
  // -- STEP 2: Adjust the example data
  // TIP: you can use copilot or similar AI-solution for this
  // by providing the entity interface to the AI and telling
  // to adjust this file to that entity interface for you.
  [
    {
      id: '1001',
      code: 'abc123',
      name: 'T-Shirt',
      description: 'MyProduct Description',
      price: 20,
      category: 'Clothing',
      quantity: 50,
      inventoryStatus: 'INSTOCK',
      rating: 4,
    },
    {
      id: '1002',
      code: 'def456',
      name: 'Jeans',
      description: 'MyProduct Description',
      price: 50,
      category: 'Clothing',
      quantity: 30,
      inventoryStatus: 'INSTOCK',
      rating: 5,
    },
    {
      id: '1003',
      code: 'ghi789',
      name: 'Shoes',
      description: 'MyProduct Description',
      price: 80,
      category: 'Footwear',
      quantity: 20,
      inventoryStatus: 'INSTOCK',
      rating: 4.5,
    },
    {
      id: '1004',
      code: 'h456wer53',
      name: 'Bracelet',
      description: 'MyProduct Description',
      price: 15,
      category: 'Accessories',
      quantity: 73,
      inventoryStatus: 'INSTOCK',
      rating: 4,
    },
    {
      id: '1005',
      code: 'jkl012',
      name: 'Hat',
      description: 'MyProduct Description',
      price: 10,
      category: 'Clothing',
      quantity: 100,
      inventoryStatus: 'INSTOCK',
      rating: 3.5,
    },
  ];

const header = 'MyProduct';

@Component({
  selector: 'angular-monorepo-my-products-test-wrapper',
  standalone: true,
  providers: [ConfirmationService],
  imports: [
    // PrimeNG
    ConfirmDialogModule,

    // Own
    EntitiesComponent,
  ],
  template: \`
    <angular-monorepo-my-products
      [data]="data"
      [messages]="messages"
      [isLoading]="isLoading"
      [header]="header"
      [noData]="noData"
      [crud]="true"
      (onDelete)="crudOperationHandler('Data deletion')"
      (onEdit)="crudOperationHandler('Data editing')"
      (onNewBeforeSubmit)="crudOperationHandler('Data submitting')"
      (onNew)="crudOperationHandler('Data submitting')"
      (onUpdate)="crudOperationHandler('Data submission')"
    >
    </angular-monorepo-my-products>
    <p-confirmDialog [style]="{ maxWidth: '450px' }"></p-confirmDialog>
  \`,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class EntityWrapperComponent extends ConfirmNotImplementedWrapperComponent {
  @Input() data: EntityInterface[] = [];
  @Input() messages: MessageInterface[] = [];
  @Input() isLoading = false;
  @Input() noData = false;
  @Input() header: string | undefined = undefined;
}

const meta: Meta<EntityWrapperComponent> = {
  component: EntityWrapperComponent,
  decorators: [applicationConfig({ ...commonAppConfig })],
  // -- STEP 3: Adjust the title for the storybook to place in the
  // correct place of the components tree
  title: 'shared-business/feature-examples/MyProduct',
};
export default meta;
type Story = StoryObj<EntityWrapperComponent>;

const expectCols = async (canvas: HTMLElement) => {
  // -- STEP 4: Adjust the columns --
  await expectText('Name', canvas);
  await expectText('Category', canvas);
  await expectText('Description', canvas);
  await expectText('Status', canvas);
  await expectText('Code', canvas);
  await expectText('Price', canvas);
  await expectText('Quantity', canvas);
  await expectText('Rating', canvas);
};

const expectValues = async (canvas: HTMLElement) => {
  // -- STEP 5: Adjust the values for the first row --
  await expectText('T-Shirt', canvas);
  await getFirstElemByText('Clothing', canvas);
  await getFirstElemByText('MyProduct Description', canvas);
  await getFirstElemByText('INSTOCK', canvas);
  await expectText('abc123', canvas);
  await getFirstElemByText('20', canvas);
  await getFirstElemByText('50', canvas);
  await getFirstElemByText('4', canvas);

  // -- STEP 6: Adjust the values for the second row --
  await expectText('Jeans', canvas);
  await getFirstElemByText('Clothing', canvas);
  await getFirstElemByText('MyProduct Description', canvas);
  await getFirstElemByText('INSTOCK', canvas);
  await expectText('def456', canvas);
  await getFirstElemByText('50', canvas);
  await getFirstElemByText('30', canvas);
  await getFirstElemByText('5', canvas);
};

export const Primary: Story = {
  args: { ...primaryTableStory.args, data: entitiesList, header },
  play: async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = getCanvas(canvasElement);

    await expectCols(canvas);
    await expectValues(canvas);

    await expectNoElem('loading', canvas);
    await expectNoElem('messages', canvas);
    await expectNoElem('no-data', canvas);
  },
};

export const Loading: Story = {
  args: {
    ...loadingStory.args,
    data: entitiesList,
    header,
  },
  play: loadingStory.play,
};

export const Error: Story = {
  args: { ...errorStory.args, header },
  play: errorStory.play,
};

export const noData: Story = {
  args: { ...noDataStory.args, header },
  play: noDataStory.play,
};
"
`;

exports[`entity-components generator should match snapshot ON any BY any 3`] = `
"import { ChangeDetectionStrategy, Component, inject } from '@angular/core';

// -- STEP 1: adjust the entity service path
import { MyProductEntityStoreService as EntityStoreService } from '@angular-monorepo/shared-business/examples';
import { MyProductsComponent as EntityComponent } from './my-products.component';

const entityHeader = 'MyProducts';

@Component({
  selector: 'angular-monorepo-my-products-smart',
  standalone: true,
  imports: [EntityComponent],
  providers: [EntityStoreService],
  template: \`
    <div data-testid="my-products-feature">
      <angular-monorepo-my-products
        [data]="entityService.entities()"
        [isLoading]="!!entityService.isLoading()"
        [messages]="entityService.messages()"
        [noData]="entityService.noData()"
        [header]="header"
        [crud]="true"
        [selectedEntity]="entityService.selectedEntity()"
        [isNewBeforeSubmitBeingEdited]="entityService.isNewEntityBeingEdited()"
        [isError]="entityService.isError()"
        (onNew)="entityService.addEntity($event)"
        (onNewBeforeSubmit)="entityService.selectUncreatedEntity()"
        (onUpdate)="entityService.updateEntity($event)"
        (onDelete)="entityService.deleteEntity($event)"
        (onEdit)="entityService.selectEntityId($event)"
        (onCancel)="entityService.resetSelectedEntity()"
      >
      </angular-monorepo-my-products>
    </div>
  \`,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class MyProductsSmartComponent {
  header = entityHeader;
  entityService = inject(EntityStoreService);
}
"
`;
